File: app/src/main/java/com/restart/spacestationtracker/PeopleInSpace.java
package com.restart.spacestationtracker;

import static com.android.volley.toolbox.HttpHeaderParser.parseCharset;

import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.airbnb.lottie.LottieAnimationView;
import com.android.volley.NetworkResponse;
import com.android.volley.ParseError;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.toolbox.HttpHeaderParser;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;
import com.restart.spacestationtracker.adapter.PeopleInSpaceAdapter;
import com.restart.spacestationtracker.data.Astronaut;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.List;

public class PeopleInSpace extends AppCompatActivity {
    private final String PEOPLE_URL = "https://corquaid.github.io/international-space-station-APIs/JSON/people-in-space.json";
    private final String BIO_URL = "https://en.wikipedia.org/w/api.php?format=json&action=query&prop=extracts&exintro=&explaintext=&titles=";
    private PeopleInSpaceAdapter mAdapter;
    private RecyclerView mRecyclerView;
    private RequestQueue mRequestQueue;
    private LottieAnimationView animation;
    private TextView errors;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.people_in_space_layout);

        mRecyclerView = findViewById(R.id.recycler);
        mRequestQueue = Volley.newRequestQueue(this);
        display_people();

        animation = findViewById(R.id.animation_view);
        errors = findViewById(R.id.errors);
    }

    private void onSuccessResult(List<Astronaut> peopleInSpace) {
        LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
        mRecyclerView.setLayoutManager(layoutManager);
        mAdapter = new PeopleInSpaceAdapter(this, peopleInSpace);
        mRecyclerView.setNestedScrollingEnabled(true);
        mAdapter.setDataSet(peopleInSpace);
        mRecyclerView.setAdapter(mAdapter);
        animation.setVisibility(View.GONE);
        errors.setVisibility(View.GONE);
    }

    private void onFailure(List<Astronaut> peopleInSpace) {
        if (peopleInSpace.isEmpty()) {
            animation.setVisibility(View.VISIBLE);
            errors.setVisibility(View.VISIBLE);
        }
    }

    /**
     * Displays a list of astronauts in a RecyclerView
     */
    private void display_people() {
        final List<Astronaut> peopleInSpace = new ArrayList<>();

        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(Request.Method.GET, PEOPLE_URL, null, astronautResponse -> {
            onSuccessResult(peopleInSpace);
            requestWiki(peopleInSpace);
        }, e -> {
            onFailure(peopleInSpace);
        }) {
            @Override
            protected Response<JSONObject> parseNetworkResponse(NetworkResponse response) {
                try {
                    String jsonString = new String(response.data, parseCharset(response.headers, PROTOCOL_CHARSET));
                    JSONObject astronautJson = new JSONObject(jsonString);

                    JSONArray astronauts = astronautJson.getJSONArray("people");

                    for (int i = 0; i < astronauts.length(); ++i) {
                        JSONObject anAstronaut = astronauts.getJSONObject(i);

                        final String name = anAstronaut.getString("name");
                        final String image = anAstronaut.getString("image");
                        final boolean isIss = anAstronaut.getBoolean("iss");
                        final String flagCode = anAstronaut.getString("flag_code").toUpperCase();
                        final int launchDate = anAstronaut.getInt("launched");
                        final String role = anAstronaut.getString("position");
                        final String location = anAstronaut.getString("spacecraft");
                        final String twitter = anAstronaut.getString("twitter");
                        final String facebook = anAstronaut.getString("facebook");
                        final String instagram = anAstronaut.getString("instagram");
                        final String wiki = anAstronaut.getString("url");

                        Astronaut storeAnAstronaut = new Astronaut(name, image, isIss, flagCode, launchDate, role, location, "", wiki, twitter, facebook, instagram);
                        peopleInSpace.add(storeAnAstronaut);
                    }

                    return Response.success(new JSONObject(jsonString), HttpHeaderParser.parseCacheHeaders(response));
                } catch (UnsupportedEncodingException | JSONException e) {
                    return Response.error(new ParseError(e));
                }
            }
        };

        mRequestQueue.add(jsonObjectRequest);
    }

    private void requestWiki(List<Astronaut> astronauts) {
        for (int i = 0; i < astronauts.size(); ++i) {
            final String wiki = astronauts.get(i).getWiki();
            final String[] wikiPage = wiki.split("/");
            final String bioUrl = BIO_URL + wikiPage[wikiPage.length - 1];

            JsonObjectRequest wikiRequest = new JsonObjectRequest(Request.Method.GET, bioUrl, null, wikiResponse -> {
                int index = 0;
                for (Astronaut astronaut: astronauts) {
                    if (astronaut.getWiki().equals(wiki)) {
                        try {
                            JSONObject pages = wikiResponse.getJSONObject("query").getJSONObject("pages");
                            JSONObject innerObject = pages.getJSONObject(pages.keys().next());
                            String bio = innerObject.getString("extract");
                            while (bio.endsWith("\n"))
                                bio = bio.substring(0, bio.length() - 2);
                            astronaut.setBio(bio);
                            mAdapter.notifyItemChanged(index);
                        } catch (JSONException e) {
                            astronaut.setBio("");
                        }
                    }
                    ++index;
                }
            }, e -> {
            }) {
                @Override
                protected Response<JSONObject> parseNetworkResponse(NetworkResponse response) {
                    try {
                        String bioJson = new String(response.data, parseCharset(response.headers, PROTOCOL_CHARSET));
                        return Response.success(new JSONObject(bioJson), HttpHeaderParser.parseCacheHeaders(response));
                    } catch (UnsupportedEncodingException | JSONException e) {
                        return Response.error(new ParseError(e));
                    }
                }
            };

            mRequestQueue.add(wikiRequest);
        }
    }
}

File: app/src/main/java/com/restart/spacestationtracker/PeopleInSpace.kt
package com.restart.spacestationtracker

import android.content.res.Configuration
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

import androidx.core.view.ViewCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

import com.airbnb.lottie.LottieAnimationView
import com.android.volley.NetworkResponse
import com.android.volley.ParseError
import com.android.volley.RequestQueue
import com.android.volley.Response
import com.android.volley.toolbox.HttpHeaderParser
import com.android.volley.toolbox.JsonObjectRequest
import com.android.volley.toolbox.Volley
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdLoader
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.MobileAds
import com.google.android.gms.ads.nativead.NativeAd
import com.restart.spacestationtracker.adapter.PeopleInSpaceAdapter
import com.restart.spacestationtracker.data.Astronaut
import com.restart.spacestationtracker.util.ON_DUTY_URL
import com.restart.spacestationtracker.util.WIKI_BIO_URL

import org.json.JSONException
import org.json.JSONObject
import java.io.UnsupportedEncodingException

import java.nio.charset.Charset

class PeopleInSpace : AppCompatActivity() {
    private lateinit var adapter: PeopleInSpaceAdapter
    private lateinit var recyclerView: RecyclerView
    private lateinit var requestQueue: RequestQueue
    private lateinit var animation: LottieAnimationView
    private lateinit var errors: TextView
    private lateinit var adLoader: AdLoader

    private val loadedAds = mutableListOf<NativeAd>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        MobileAds.initialize(this) {}

        WindowCompat.setDecorFitsSystemWindows(window, false)
        val currentNightMode = resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK
        val isNightMode = currentNightMode == Configuration.UI_MODE_NIGHT_YES

        val windowInsetsController = WindowCompat.getInsetsController(window, window.decorView)
        windowInsetsController.isAppearanceLightStatusBars = !isNightMode
        setContentView(R.layout.people_in_space_layout)

        val mainLayout = findViewById<View>(R.id.main_layout)
        recyclerView = findViewById(R.id.recycler)

        ViewCompat.setOnApplyWindowInsetsListener(mainLayout) { _, windowInsets ->
            val insets = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars())
            recyclerView.setPadding(
                recyclerView.paddingLeft,
                insets.top,
                recyclerView.paddingRight,
                insets.bottom
            )
            WindowInsetsCompat.CONSUMED
        }

        recyclerView.clipToPadding = false

        requestQueue = Volley.newRequestQueue(this)
        displayPeople()

        animation = findViewById(R.id.animation_view)
        errors = findViewById(R.id.errors)
    }

    private fun onSuccessResult(peopleInSpace: MutableList<Astronaut>) {
        loadAndInsertAds(peopleInSpace)
        requestWiki(peopleInSpace)
    }

    private fun loadAndInsertAds(originalList: List<Astronaut>) {
        if (originalList.isEmpty()) {
            setupRecyclerView(originalList)
            return
        }

        val adsToRequest = originalList.size / ITEMS_PER_AD
        if (adsToRequest <= 0) {
            setupRecyclerView(originalList)
            return
        }

        loadedAds.clear()

        adLoader = AdLoader.Builder(this, getString(R.string.on_duty_native_ad_unit_id))
            .forNativeAd { nativeAd ->
                loadedAds.add(nativeAd)
            }
            .withAdListener(object : AdListener() {
                override fun onAdFailedToLoad(loadAdError: LoadAdError) {
                    super.onAdFailedToLoad(loadAdError)
                    Log.e(TAG, "Native ad failed to load: ${loadAdError.message}")
                    if (!adLoader.isLoading) {
                        insertAdsInList(originalList)
                    }
                }

                override fun onAdLoaded() {
                    super.onAdLoaded()
                    if (!adLoader.isLoading) {
                        insertAdsInList(originalList)
                    }
                }
            })
            .build()

        adLoader.loadAds(AdRequest.Builder().build(), adsToRequest)
    }

    private fun insertAdsInList(originalList: List<Astronaut>) {
        if (loadedAds.isEmpty()) {
            setupRecyclerView(originalList)
            return
        }
        val combinedList = mutableListOf<Any>()
        var adIndex = 0
        originalList.forEachIndexed { index, astronaut ->
            combinedList.add(astronaut)
            if ((index + 1) % ITEMS_PER_AD == 0 && adIndex < loadedAds.size) {
                combinedList.add(loadedAds[adIndex])
                adIndex++
            }
        }
        setupRecyclerView(combinedList)
    }

    private fun setupRecyclerView(list: List<Any>) {
        adapter = PeopleInSpaceAdapter(this, list)
        recyclerView.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        recyclerView.isNestedScrollingEnabled = true
        recyclerView.adapter = adapter
        animation.visibility = View.GONE
        errors.visibility = View.GONE
    }

    private fun onFailure(peopleInSpace: MutableList<Astronaut>) {
        if (peopleInSpace.isEmpty()) {
            animation.visibility = View.VISIBLE
            errors.visibility = View.VISIBLE
        }
    }

    private fun displayPeople() {
        val peopleInSpace = mutableListOf<Astronaut>()

        val jsonObjectRequest = object : JsonObjectRequest(
            Method.GET, ON_DUTY_URL, null,
            {
                onSuccessResult(peopleInSpace)
            },
            {
                onFailure(peopleInSpace)
            }
        ) {
            override fun parseNetworkResponse(response: NetworkResponse): Response<JSONObject> {
                return try {
                    val jsonString = String(
                        response.data,
                        Charset.forName(
                            HttpHeaderParser.parseCharset(
                                response.headers,
                                PROTOCOL_CHARSET
                            )
                        )
                    )
                    val astronautJson = JSONObject(jsonString)
                    val astronauts = astronautJson.getJSONArray("people")

                    for (i in 0 until astronauts.length()) {
                        val anAstronaut = astronauts.getJSONObject(i)
                        val astronaut = Astronaut(
                            name = anAstronaut.getString("name"),
                            image = anAstronaut.getString("image"),
                            isIss = anAstronaut.getBoolean("iss"),
                            flagCode = anAstronaut.getString("flag_code").uppercase(),
                            launchDate = anAstronaut.getInt("launched"),
                            role = anAstronaut.getString("position"),
                            location = anAstronaut.getString("spacecraft"),
                            wiki = anAstronaut.getString("url"),
                            twitter = anAstronaut.getString("twitter"),
                            facebook = anAstronaut.getString("facebook"),
                            instagram = anAstronaut.getString("instagram"),
                            bio = ""
                        )
                        peopleInSpace.add(astronaut)
                    }
                    Response.success(
                        JSONObject(jsonString),
                        HttpHeaderParser.parseCacheHeaders(response)
                    )
                } catch (e: Exception) {
                    when (e) {
                        is UnsupportedEncodingException, is JSONException -> Response.error(
                            ParseError(e)
                        )

                        else -> throw e
                    }
                }
            }
        }

        requestQueue.add(jsonObjectRequest)
    }

    private fun requestWiki(astronauts: MutableList<Astronaut>) {
        astronauts.forEach { astronaut ->
            val wikiPage = astronaut.wiki.split("/")
            val bioUrl = "$WIKI_BIO_URL${wikiPage.last()}"

            val wikiRequest = object : JsonObjectRequest(
                Method.GET, bioUrl, null,
                { response ->
                    try {
                        val pages = response.getJSONObject("query").getJSONObject("pages")
                        val innerObject = pages.getJSONObject(pages.keys().next())
                        var bio = innerObject.getString("extract")
                        while (bio.endsWith("\n")) {
                            bio = bio.substring(0, bio.length - 2)
                        }
                        astronaut.bio = bio
                        if (::adapter.isInitialized) {
                            val position = adapter.findPositionOf(astronaut)
                            if (position != -1) {
                                adapter.notifyItemChanged(position)
                            }
                        }
                    } catch (e: JSONException) {
                        astronaut.bio = ""
                    }
                },
                { }
            ) {
                override fun parseNetworkResponse(response: NetworkResponse): Response<JSONObject> {
                    return try {
                        val bioJson = String(
                            response.data,
                            Charset.forName(
                                HttpHeaderParser.parseCharset(
                                    response.headers,
                                    PROTOCOL_CHARSET
                                )
                            )
                        )
                        Response.success(
                            JSONObject(bioJson),
                            HttpHeaderParser.parseCacheHeaders(response)
                        )
                    } catch (e: Exception) {
                        when (e) {
                            is UnsupportedEncodingException, is JSONException -> Response.error(
                                ParseError(e)
                            )

                            else -> throw e
                        }
                    }
                }
            }
            requestQueue.add(wikiRequest)
        }
    }

    companion object {
        private const val TAG = "PeopleInSpace"
        private const val ITEMS_PER_AD = 4
    }
}